ferry.go

```diff
@@ -205,56 +283,16 @@ func (f *Ferry) Initialize() (err error) {
 		f.Throttler = &PauserThrottler{}
 	}
 
-	f.StateTracker = NewStateTracker(f.DataIterationConcurrency * 10)
-
-	f.BinlogStreamer = &BinlogStreamer{
-		Db:           f.SourceDB,
-		Config:       f.Config,
-		ErrorHandler: f.ErrorHandler,
-		Filter:       f.CopyFilter,
-	}
-	err = f.BinlogStreamer.Initialize()
-	if err != nil {
-		return err
-	}
-
-	f.BinlogWriter = &BinlogWriter{
-		DB:               f.TargetDB,
-		DatabaseRewrites: f.Config.DatabaseRewrites,
-		TableRewrites:    f.Config.TableRewrites,
-		Throttler:        f.Throttler,
-
-		BatchSize:    f.Config.BinlogEventBatchSize,
-		WriteRetries: f.Config.DBWriteRetries,
-
-		ErrorHandler: f.ErrorHandler,
-		StateTracker: f.StateTracker,
-	}
-
-	err = f.BinlogWriter.Initialize()
-	if err != nil {
-		return err
-	}
-
-	f.DataIterator, err = f.newDataIterator()
-	if err != nil {
-		return err
-	}
-
-	f.BatchWriter = &BatchWriter{
-		DB:           f.TargetDB,
-		StateTracker: f.StateTracker,
-
-		DatabaseRewrites: f.Config.DatabaseRewrites,
-		TableRewrites:    f.Config.TableRewrites,
-
-		WriteRetries: f.Config.DBWriteRetries,
+	if f.StateToResumeFrom == nil {
+		f.StateTracker = NewStateTracker(f.DataIterationConcurrency * 10)
```

Comment:
What's this magical number 10 here?

Reply:
There's a ring inside the StateTracker that tracks the speed at which we are copying data. The argument here specifies the ring's size. It should be a multiple of the number of goroutines that could be writing to the ring. 10 is just chosen as a conservative estimate.